### PR TITLE
pari: update to 2.13.2.

### PR DESCRIPTION
**Pull request by tornaria**
- enable pthreads build
- switch gp binary to static link libpari; this offsets the slowdown due to pthreads
- use -flto and -fno-semantic-interposition which improve speed a little bit more
- ship the static library, so external programs can static link
- change of maintainer agreed by sgn
- revbump giac and qcas

**Rationale:**
- pthreads gives an interesting functionality, and recent versions of giac require pari with tls; sagemath requires pari with pthreads.
- The cost for enabling pthreads is quite high, on the order of 30-40% for some tests; `-flto` saves maybe 1-2% at most, `-fno-semantic-interposition` saves 5-7%, but we are still about 25-30% behind.
- On the other hand, switching to static binary yields some savings on the order of 5-10%
- More important: the cost of enabling pthread with -flto for a static binary is only about 5-10%
- In summary: enabling pthread while switching to static is roughly the same speed (even maybe 5-10% better overall) as current binary, with an interesting feature (esp. if one has many cores) and apparently better compatibility.
- Shipping static library as well is easy and allows for external code linking to the static library.

@sgn: if you agree, I am willing to take over maintainership. I have used pari for 25+ years, and I use it almost daily.

#### Have the results of the proposed changes been tested?
- [x] I use the packages affected by the proposed changes on a regular basis and confirm this PR works for me

### CODE REVIEW

**Comment by sgn** [comment]
The package is yours.

**Comment by ericonr** [comment]
> switch gp binary to static link libpari; this offsets the slowdown due to pthreads

Can you try the `-fno-semantic-interposition` compiler argument? Python has started using it for libpython for similar performance gains. I think it might allow us to keep the space gain of dynamic linking and the speed gain of static linking.

**Comment by tornaria** [comment]
> Can you try the `-fno-semantic-interposition` compiler argument? Python has started using it for libpython for similar performance gains. I think it might allow us to keep the space gain of dynamic linking and the speed gain of static linking.

A quick informal test shows that while adding that option improves the dynamic binary a bit, it's not close to recovering what is lost from pthreads. Running the script below with the current version in the repos (no pthreads, no CFLAGS):
```
$ gp -q regress-best5.gp
4470 4491 4500 4532 4509   best=4470
```

Now with the new version, enabling pthreads:
- CFLAGS=""
gp-dyn: 5970 6000 6006 6010 6020   best=5970
gp-sta: 4219 4280 4210 4154 4203   best=4154

- CFLAGS="-flto"
gp-dyn: 6226 6168 6145 6196 6095   best=6095
gp-sta: 4166 4045 4028 4010 4013   best=4010

- CFLAGS="-flto -fno-semantic-interposition"
gp-dyn: 5669 5680 5702 5686 5628   best=5628
gp-sta: 3900 3934 4050 3921 3963   best=3900

So for this particular test, there's a 40% loss by going dynamic with `-fno-semantic-interposition` instead of 50% loss without. 

Note that we are NOT static linking system libraries. It's only `libpari` that is statically linked into the `gp` instead of being dynamically linked.

---




- regress-best_of_5.gp
```
sigmatwist1(n,k)=
{
   if(denominator(n)>1||n<=0,return(0));
   n>>=valuation(n,2);
   sumdiv(n,d,if(d%4==1,d^k,-d^k));
}

sumtwist1(k,m,N)=
{
   sigmatwist1(m/N,k-1)+2*sum(s=1,sqrtint(m),sigmatwist1((m-s^2)/N,k-1));
}

a={vector(5,i,
  gettime();
  sumtwist1(3,(10^12+4)/4,1);
  t=gettime();
  print1(t," ");
  t)};
  print("  best=",vecmin(a));
\q
```

**Comment by tornaria** [comment]
Failures are due to a dependant package (giac) which is affected since the soname changed (because of the switch to pthreads).

I'll deal with that, and maybe take the chance to upgrade giac.

**Comment by ericonr** [comment]
I think linking it in statically is valid. Quite unfortunate that the flag wasn't enough, but it might be more pronounced on python than on gp, especially if the binary makes lots of calls to the library instead of a single call and now the rest runs inside the library.

**Comment by tornaria** [comment]
All tests are ok, except on aarch64 and aarch64-musl it's trying to build qt5, which is a dependency of qcas.

I think pari and giac are built ok before that, so I'd take this as check ok for all architectures.

I didn't change qcas at all, only revbumped and fix a xlint in the short_desc. As a matter of fact, the current version of qcas in the repos is already broken before me changing pari due to a missing symbol; maybe no one uses qcas (I don't). Anyway after this PR it works at least on x86-64 so it's definitely an improvement.

File: srcpkgs/giac/patches/pari_2_13.patch [hunk 1]
```diff
@@ -0,0 +1,13 @@
+change in order to pass chk_fhan4 with pari-2.13
+
+--- a/src/pari.cc	2019-11-21 13:40:55.000000000 +0000
++++ b/src/pari.cc	2021-08-11 20:51:39.243502400 +0000
+@@ -88,7 +88,7 @@
+   static map<string,entree *> pari_function_table;
+   static void do_giac_pari_init(long maxprime){
+ #ifdef PARI_DYNAMIC_STACK
+-    long pari_mem_size=512000;
++    long pari_mem_size=1024000;
```
Comment (ericonr) [inline srcpkgs/giac/patches/pari_2_13.patch:10]: Please try to upstream this ;)

File: srcpkgs/pari/template [hunk 2]
```diff
@@ -1,24 +1,29 @@
...
+version=2.13.2
 revision=1
 build_style=configure
 build_helper=qemu
 configure_script=./Configure
-configure_args="--prefix=/usr"
+configure_args="--prefix=/usr $(vopt_if pthreads --mt=pthread)"
 make_build_target=all
 make_check_target=test-all
+make_install_target="install install-bin-sta install-lib-sta"
```
Comment (ericonr) [inline srcpkgs/pari/template:11]: It seems to be running a compiler in the install steps, probably because of `install-bin-sta` and `install-lib-sta`. Maybe add targets to `make_build_target` so everything happens in `do_build`?

File: common/shlibs [hunk 3]
```diff
@@ -3980,7 +3980,7 @@ libopenaptx.so.0 libopenaptx-0.2.0_1
 libsimavr.so.1 simavr-1.6_2
 libsimavrparts.so.1 simavr-1.6_2
 libsword-1.8.1.so libsword-1.8.1_6
-libpari-gmp.so.7 pari-2.13.0_1
+libpari-gmp-tls.so.7 pari-2.13.2_1
```
Comment (ericonr) [inline common/shlibs:3983]: Don't you kind of need both versions here, since it can be built without `pthreads`?

Comment (ericonr) [reply common/shlibs:3983]: Alternatively, completely remove the build option and force pthreads always :p

File: srcpkgs/giac/patches/pari_2_13.patch (hunk 1 above)
Comment (tornaria) [reply srcpkgs/giac/patches/pari_2_13.patch:10]: This is already in upstream. I'm working on an update of giac to 1.7.0-25 which won't need this patch.

File: srcpkgs/pari/template (hunk 2 above)
Comment (tornaria) [reply srcpkgs/pari/template:11]: I tried, but they don't seem to have compile targets for that, only install targets. If I'm not mistaken it's not compiling anything, just linking  (it takes a long time because of -flto).

If this is really important, I think I can easily patch the makefile to add compile targets for the static binary and library.

File: common/shlibs (hunk 3 above)
Comment (tornaria) [reply common/shlibs:3983]: You are right, I will fix that.

File: srcpkgs/pari/template (hunk 2 above)
Comment (ericonr) [reply srcpkgs/pari/template:11]: Not super important, but would be nice to achieve upstream, if you can :)

Since it's using a configure build style, we don't need to pass args to make and it has a smaller chance of breaking, but it would be nice to confine building to `do_build`.

From what I see in https://pari.math.u-bordeaux.fr/cgi-bin/gitweb.cgi?p=pari.git;a=blob;f=config/Makefile.SH;h=ca8bbf31a718c39abc6450c38aa176f6b157738f;hb=HEAD , you might be able to do `make_build_target="all lib-sta lib-dyn gp-sta"`.

Comment (tornaria) [reply srcpkgs/pari/template]: I had to patch the makefile so I don't to need to change `do_build()`. I will try to upstream that patch.

File: srcpkgs/giac/patches/pari_2_13.patch (hunk 1 above)
Comment (tornaria) [reply srcpkgs/giac/patches/pari_2_13.patch:13]: Added a note in the patch. We'll be removing that patch soon.

**Comment by tornaria** [comment]
If all checks pass, then it's ready. Thanks a lot for your careful and constructive review.

**Comment by ericonr** [comment]
aarch64 is just a tad stuck :p 

Thanks!
